heat_pump_adoption_modelling/pipeline/preprocessing/feature_engineering.py

```diff
@@ -0,0 +1,588 @@
+"""
+
+"""
+Adding new features to EPC dataset.
+"""
+
+# ----------------------------------------------------------------------------------
+
+# Import
+import pandas as pd
+import re
+from hashlib import md5
+
+from heat_pump_adoption_modelling import getters
+from heat_pump_adoption_modelling.pipeline.preprocessing import data_cleaning
+
+# ----------------------------------------------------------------------------------
+
+
+def short_hash(text):
+    """Generate a unique short hash for given string.
+
+    Parameters
+    ----------
+    text: str
+        String for which to get ID.
+
+    Return
+    ---------
+
+    short_code: int
+        Unique ID."""
+
+    hx_code = md5(text.encode()).hexdigest()
+    int_code = int(hx_code, 16)
+    short_code = str(int_code)[:16]
+    return int(short_code)
+
+
+def get_unique_building_ID(df):
+    """Add unique building ID column to dataframe.
+
+    Parameters
+    ----------
+    text: str
+        String for which to get ID.
+
+    Return
+    ---------
+
+    short_code: int
+        Unique ID."""
+
+    # Remove samples with no address
+    df.dropna(subset=["ADDRESS1"], inplace=True)
+
+    # Create unique address and building ID
+    df["UNIQUE_ADDRESS"] = df["ADDRESS1"] + df["POSTCODE"]
+    df["BUILDING_ID"] = df["UNIQUE_ADDRESS"].apply(short_hash)
+
+    return df
+
+
+def get_new_EPC_rating_features(df):
+    """Get new EPC rating features related to EPC ratings.
+
+        CURR_ENERGY_RATING_NUM: EPC rating representeed as number
+        high number = high rating.
+
+        ENERGY_RATING_CAT: EPC range category.
+        A-B, C-D or E-G
+
+        DIFF_POT_ENERGY_RATING: Difference potential and current
+        energy rating.
+
+
+    Parameters
+    ----------
+    df : pandas.Dataframe
+        EPC dataframe.
+
+    Return
+    ---------
+    df : pandas.DateFrame
+        Updated EPC dataframe with new EPC rating features."""
+
+    # EPC rating dict
+    rating_dict = {
+        "A": 7,
+        "B": 6,
+        "C": 5,
+        "D": 4,
+        "E": 3,
+        "F": 2,
+        "G": 1,
+        "H": 0,
+        "INVALID!": 0,
+        "unknown": 0,
+    }
+
+    # EPC range cat dict
+    EPC_cat_dict = {
+        "A": "A-B",
+        "B": "A-B",
+        "C": "C-D",
+        "D": "C-D",
+        "E": "E-G",
+        "F": "E-G",
+        "G": "E-G",
+    }
+
+    # EPC rating in number instead of letter
+    df["CURR_ENERGY_RATING_NUM"] = df.CURRENT_ENERGY_RATING.map(rating_dict)
+
+    # EPC rating in category (A-B, C-D or E-G)
+    df["ENERGY_RATING_CAT"] = df.CURRENT_ENERGY_RATING.map(EPC_cat_dict)
+
+    # Numerical difference between current and potential energy rating (A-G)
+    df["DIFF_POT_ENERGY_RATING"] = (
+        df.POTENTIAL_ENERGY_RATING.map(rating_dict) - df["CURR_ENERGY_RATING_NUM"]
+    )
+
+    # Set DIFF_POT_ENERGY_RATING to 0.0
+    # if substraction yielded value below 0.0 due to input error (few cases).
+    df = df[df.DIFF_POT_ENERGY_RATING >= 0.0]
```

Comment:
Right!
I decided to remove those samples at some point but forgot to adjust the comment.

Since I anyway had to correct the handling of unknown EPC rating (see above), I now catch unreasonable DIFF_POT_ENERGY_RATINGs (values below 0 and above 7) and set them to "unknown" as it suggests one of the EPC ratings was "unknown". 

Thanks for bringing this to my attention!